buffer: add Buffer compare by offset

jasnell opened the pull request:
### Pull Request check-list

- [x] Does `make -j8 test` (UNIX) or `vcbuild test nosign` (Windows) pass with
  this change (including linting)?
- [x] Is the commit message formatted according to [CONTRIBUTING.md][0]?
- [x] If this change fixes a bug (or a performance problem), is a regression
  test (or a benchmark) included?
- [x] Is a documentation update included (if this change modifies
  existing APIs, or introduces new ones)?

### Affected core subsystem(s)

buffer

### Description of change

Adds additional `offset`, `length` and `thisOffset` arguments
to `Buffer.prototype.compare` to allow comparison of sub-ranges
of two Buffers without requiring Buffer.prototype.slice()

Fixes: https://github.com/nodejs/node/issues/521
/cc @trevnorris @rootslab @jorangreef

mscdex commented on doc/api/buffer.markdown:
````diff
@@ -670,9 +670,15 @@ console.log(buf.toString('ascii'));
 ```
 
-### buf.compare(otherBuffer)
+### buf.compare(otherBuffer[, offset[, length[, thisOffset]]])
````
Perhaps the order should be tweaked a bit here and have `length` and `thisOffset` swapped to match more closely with `buffer.copy()` which also deals with two buffers?

I also think `offset` might be better renamed to something like `otherOffset`.

trevnorris replied:
Possibly follow suit of `copy()` since it follows similar parameters.

mscdex commented on lib/buffer.js:
```diff
@@ -290,7 +290,6 @@ Buffer.compare = function compare(a, b) {
   return binding.compare(a, b);
 };
 
-
```
Unintentional whitespace removal?

mscdex commented on lib/buffer.js:
```diff
@@ -534,14 +533,33 @@ Buffer.prototype.inspect = function inspect() {
   return '<' + this.constructor.name + ' ' + str + '>';
 };
 
-
```
Ditto?

jasnell commented:
@trevnorris @mscdex ... updated! PTAL

The signature has been changed to: `buf.compare(other[, otherStart[, thisStart[, otherEnd[, thisEnd]]]])`

mscdex commented:
Couldn't we just have one end value, `thisEnd`, or just keep `length` instead of the end parameters? If not, I would prefer seeing `buf.compare(other[, otherStart[, otherEnd[, thisStart[, thisEnd]]]])`.

jasnell commented:
Personally I'd prefer to keep `length`. I changed it to this to align with the signature for `buf.copy()`

jasnell commented:
That said, having it as just `length` limits you to always comparing equal length ranges which is not always desirable.

jasnell commented:
@mscdex ... updated ... I swapped the order of the arguments but kept separate `otherEnd` and `thisEnd`. 

The signature has been changed to: `buf.compare(other[, otherStart[, otherEnd[, thisStart[, thisEnd]]]])`

jasnell commented:
Currently, `thisEnd` defaults to `this.byteLength`, however, we could make it so that if `thisEnd` is not specified, it defaults to `Math.min(this.byteLength, thisStart + (otherEnd - otherStart)))`

jasnell commented:
doing so causes surprising behavior tho:

```js
const a = Buffer.from([1,2,3,4,5,6,7,8,9]);
const b = Buffer.from([5,6,7,8,9,1,2,3,4]);
console.log(a.compare(b, 5, 9));
  // Prints 0
```

mscdex commented:
I don't think it's that much of a surprising behavior considering that the default for `sourceStart` for `.copy()` is also 0.

jasnell commented:
the surprising behavior is that the subset of b is not compared to the full content of a. If I did `Buffer.compare(a, b.slice(5, 9))` I would get a different result.

trevnorris commented on doc/api/buffer.markdown:
```diff
@@ -670,9 +670,19 @@ console.log(buf.toString('ascii'));
+  default = `otherBuffer.byteLength`.
+* `thisStart` {Integer} The offset within `buf` at which to begin comparison.
+  Ignored when `otherStart` is undefined. default = `0`
+* `thisEnd` {Integer} The offset within `buf` at which to end comparison.
```
Can we use `source`/`target` to mirror `copy()`? Think it will have less mental impact.

jasnell replied:
we could but `source`/`target` make a lot more sense in `copy` than they do in `compare`.

trevnorris commented:
@jasnell One documentation comment, but implementation LGTM.

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/2069/

jasnell commented:
@mscdex ... any further thoughts on this one?

mscdex commented on benchmark/buffers/buffer-compare-offset.js:
```diff
@@ -0,0 +1,44 @@
+
+  switch (conf.method) {
+    case 'slice':
+      bench.start();
```
What about moving the `bench.start()`/`bench.end()` to be closer to the actual loop in the respective functions?

mscdex commented on benchmark/buffers/buffer-compare-offset.js:
```diff
@@ -0,0 +1,44 @@
+
+function compareUsingSlice(b0, b1, size, iter) {
+  const len = size >> 2;
+  for (var i = 0; i < iter; i++)
```
In here we should probably force optimization of `Buffer.compare()`, `b0.slice()`, and `b1.slice()`. See the `path` benchmarks on how to do this.

mscdex commented on benchmark/buffers/buffer-compare-offset.js:
```diff
@@ -0,0 +1,44 @@
+});
+
+function compareUsingSlice(b0, b1, size, iter) {
+  const len = size >> 2;
```
Before the loop here (and before `bench.start()`) we should probably force optimization of `Buffer.compare()`, `b0.slice()`, and `b1.slice()`. See the `path` benchmarks on how to do this.

mscdex commented on benchmark/buffers/buffer-compare-offset.js:
```diff
@@ -0,0 +1,44 @@
+}
+
+function compareUsingOffset(b0, b1, size, iter) {
+  const len = size >> 2;
```
Ditto here, `b0.compare()` should be forced to optimize before starting the loop and benchmark.

jasnell replied:
interestingly.. forcing the optimization actually *slows* `b0.compare` down by a significant amount.

mscdex commented on doc/api/buffer.markdown:
```diff
@@ -670,18 +670,26 @@ console.log(buf.toString('ascii'));
+* `targetStart` {Integer} The offset within `target` at which to begin
+  comparison. default = `0`.
+* `targetEnd` {Integer} The offset with `target` at which to end comparison.
+  Ignored with `otherStart` is undefined. default = `target.byteLength`.
```
s/with/when/

mscdex replied:
Also, perhaps 'undefined' should be surrounded by backticks?

mscdex commented on doc/api/buffer.markdown:
```diff
@@ -703,6 +711,22 @@ console.log(buf2.compare(buf3));
+console.log(buf1.compare(buf2, 0, 6, 4));
+  // Prints: -1
+console.log(buf1.compare(buf2, 5, 6, 5));
+ // Prints: 1
```
nit: comment alignment on this line doesn't match the previous comments

mscdex commented on lib/buffer.js:
```diff
@@ -535,10 +535,35 @@ Buffer.prototype.inspect = function inspect() {
 
+  if (typeof start !== 'undefined' &&
+      (typeof start !== 'number' ||
+       +start !== (start >> 0)))
```
I thought we were adding braces to conditionals that spanned multiple lines themselves, despite the number of lines in their actual blocks?

jasnell replied:
We are, just missed it because linter doesn't appear to be picking them up.

mscdex commented on lib/buffer.js:
```diff
@@ -535,10 +535,35 @@ Buffer.prototype.inspect = function inspect() {
+
+  if (typeof start !== 'undefined') {
+    thisStart >>= 0;
+    start >>= 0;
```
This will silently convert some non-`undefined` values to 0, such as `null`, `false`, and `'0'`. Should this be supported/expected?

mscdex commented on lib/buffer.js:
```diff
@@ -535,10 +535,35 @@ Buffer.prototype.inspect = function inspect() {
+  if (typeof start !== 'undefined') {
+    thisStart >>= 0;
+    start >>= 0;
+    thisEnd = (isNaN(thisEnd) ? this.byteLength : thisEnd >> 0);
```
Is it even possible for `isNaN()` to ever return `true`, due to the nature of the value checks performed earlier in this function?

trevnorris replied:
undefined is enough. Everything else counts as a passed argument and coerced if applicable.

mscdex commented:
@jasnell Just a few more comments.

jasnell commented:
@trevnorris @mscdex ... updated and rebased. PTAL!

mscdex commented:
My question about the usage of `>>= 0` still stands. Do we want to add more type checks or ?

jasnell commented:
If anything, the type checks that are already there make the `>>= 0` checks unnecessary. We're already validating that the inputs are numbers and that they're proper integers, and the native compare does a proper bounds check. I added the `>>= 0` out of an abundance of caution. I've removed them now.

trevnorris commented on benchmark/buffers/buffer-compare-offset.js:
```diff
@@ -0,0 +1,50 @@
+  var i;
+  bench.start();
+  for (i = 0; i < iter; i++)
+    Buffer.compare(b0.slice(1, len), b1.slice(1, len));
```
Mind throwing this into it's own function? Easier to view optimized output when run from own function.

trevnorris commented on benchmark/buffers/buffer-compare-offset.js:
```diff
@@ -0,0 +1,50 @@
+      compareUsingSlice : compareUsingOffset;
+  method(Buffer.alloc(size, 'a'),
+         Buffer.alloc(size, 'b'),
+         size >> 2,
```
Do you mean to divide by 2 or 4?

jasnell replied:
heh, 2 ... instance of thinking one thing and typing another ;-)

trevnorris commented on doc/api/buffer.markdown:
```diff
@@ -675,18 +675,26 @@ console.log(buf.toString('ascii'));
-* `-1` is returned if `otherBuffer` should come *after* `buf` when sorted.
+* `0` is returned if `target` is the same as `buf`
+* `1` is returned if `target` should come *before* `buf` when sorted.
+* `-1` is returned if `target` should come *after* `buf` when sorted.
```
Should document whether this will throw if, say, `sourceEnd > buf.length`, or if it will simply return `-1`.

trevnorris commented on src/node_buffer.cc:
```diff
@@ -870,6 +870,62 @@ void ByteLengthUtf8(const FunctionCallbackInfo<Value> &args) {
+  CHECK_NOT_OOB(ParseArrayIndex(args[5], ts_obj_length, &source_end));
+
+  if (target_start >= target_end || source_start >= source_end)
+    return args.GetReturnValue().Set(0);
```
I would expect this case to operate the same as `Buffer.compare(Buffer(0), Buffer(2))`. Which returns `-1`.

trevnorris replied:
I'd expect:
```cpp
if (target_start >= target_end && source_start >= source_end)
  return args.GetReturnValue().Set(0);
else if (source_start >= source_end)
  return args.GetReturnValue().Set(-1);
else if (target_start >= target_end)
  return args.GetReturnValue().Set(1);
```

trevnorris commented on src/node_buffer.cc:
```diff
@@ -870,6 +870,62 @@ void ByteLengthUtf8(const FunctionCallbackInfo<Value> &args) {
+  CHECK_NOT_OOB(ParseArrayIndex(args[2], 0, &target_start));
+  CHECK_NOT_OOB(ParseArrayIndex(args[3], 0, &source_start));
+  CHECK_NOT_OOB(ParseArrayIndex(args[4], target_length, &target_end));
+  CHECK_NOT_OOB(ParseArrayIndex(args[5], ts_obj_length, &source_end));
```
curse me for using `CHECK` in the macro name.

trevnorris commented on lib/buffer.js:
```diff
@@ -539,11 +539,29 @@ Buffer.prototype.inspect = function inspect() {
+function assertInteger(val, name) {
+  if (typeof val !== 'undefined' &&
+      (typeof val !== 'number' ||
+       +val !== (val >> 0))) {
```
`node::ParseArrayIndex()` uses `Uint32Value()`. So we should use uint32 checks in JS to prevent UB. As this stands, if I pass `-1` it'll be coerced to `4294967295` when it hits native. Also true because you're assigning the value to a `size_t`.

trevnorris replied:
Easiest way to do this is `val != (val >>> 0)`. Though this also doesn't follow other Buffer APIs. We coerce almost everywhere else.

trevnorris commented on lib/buffer.js:
```diff
@@ -539,11 +539,29 @@ Buffer.prototype.inspect = function inspect() {
+  assertInteger(start, 'targetStart');
+  assertInteger(thisEnd, 'sourceEnd');
+  assertInteger(end, 'targetEnd');
+  assertInteger(thisStart, 'sourceStart');
```
Given what I said above, I'd suggest using checks closer to what's in `slowToString()`. We went through a lot of discussion making sure those checks/coersions to make sure those covered every case (ref ec836547).

That being said, I'd do it like so:
```js
if (thisEnd === undefined || thisEnd > this.length)
  thisEnd = this.length;
if (thisStart === undefined || thisStart < 0)
  thisStart = 0;
if (end === undefined || end > b.length)
  end = b.length;
if (start === undefined || start < 0)
  start = 0;

if (thisStart >= thisEnd && start >= end)
  return 0;
if (thisStart >= thisEnd)
  return -1;
if (start >= end)
return 1;

// Range comparisons above handle the case where end
// and thisEnd are < 0. So can do uint32 coercion without
// wrap around. Properly handles coercion of falsey/NaN
// values.
start >>>= 0;
end >>>= 0;
thisStart >>>= 0;
thisEnd >>>= 0;

return binding.compareOffset(this, b, start, thisStart, end, thisEnd);
```

trevnorris replied:
My above handles the specific case to always coerce ranges for `toString()` and `slice()` (which aren't allowed to throw). Though in the past I've tried to enforce more strict range checks. If we want to proceed down that path (not sure of this now that we're using typed arrays, which seem like they don't like to throw) then it should look something like (going to use new syntax. benchmark shows minor performance lose that should be mostly recovered in v5):
```js
Buffer.prototype.compare = function compare(b,
                                            start = 0,
                                            end = b.length,
                                            thisStart = 0,
                                            thisEnd = this.length) {
  if (thisStart >= thisEnd && start >= end)
    return 0;
  if (thisStart >= thisEnd)
    return -1;
  if (start >= end)
    return 1;

  if (start < 0 ||
      end > b.length ||
      thisStart < 0 ||
      thisEnd > this.length) {
    throw new RangeError('out of range index');
  }

  // Range comparisons above handle the case where end
  // and thisEnd are < 0. So can do uint32 coercion without
  // wrap around. Properly handles coercion of falsey/NaN
  // values.
  start >>>= 0;
  end >>>= 0;
  thisStart >>>= 0;
  thisEnd >>>= 0;

  return binding.compareOffset(this, b, start, thisStart, end, thisEnd);
}
```

trevnorris replied:
If you want to experiment moving all of this to native for performance, be my guest.

jasnell replied:
are argument defaults being optimized in v5? if so, this is goodness.

trevnorris commented on test/parallel/test-buffer-compare-offset.js:
```diff
@@ -0,0 +1,73 @@
+assert.throws(() => a.compare(b, 0, -1), oor);
+assert.throws(() => a.compare(b, 0, 0, -1), oor);
+assert.throws(() => a.compare(b, 0, 0, 0, -1), oor);
+assert.throws(() => a.compare(), arg_re);
```
Can we add tests passing values like `'0xff'`, `null`, `Infinity` and even `{ valueOf: () => 5 }`?

trevnorris commented:
@jasnell excuse all the comments. finished review.

jasnell commented:
No worries on the comments :-) I'll get back to this either tonight or on friday.

jasnell commented:
@trevnorris ... updated to address your feedback. PTAL!

Note: I didn't go with the new arg default syntax because v8 is deoptimizing way too much.

jasnell commented:
Commits squashed and rebased. New CI: https://ci.nodejs.org/job/node-test-pull-request/2143/

trevnorris commented on lib/buffer.js:
```diff
@@ -537,15 +537,44 @@ Buffer.prototype.inspect = function inspect() {
+  if (typeof thisStart === 'undefined')
+    thisStart = 0;
+  if (typeof thisEnd === 'undefined')
+    thisEnd = this.length;
```
Nice. For reference, this should operate the same way as typed arrays:
```js
new Uint8Array(new ArrayBuffer(8), undefined, 5).byteOffset === 0
new Uint8Array(new ArrayBuffer(8), false, 5).byteOffset === 0
new Uint8Array(new ArrayBuffer(8), true, 5).byteOffset === 1
new Uint8Array(new ArrayBuffer(8), 0, undefined).byteLength === 0
new Uint8Array(new ArrayBuffer(8), 0, null).byteLength === 0
new Uint8Array(new ArrayBuffer(8), 0, true).byteLength === 1
```
Do you believe these existing tests covers this?

I may recommend dropping the `typeof`. Not important enough for me to block, but not sure there's a need for the extra keyword.

jasnell replied:
I believe we're covered but I'll double check. And +1 on dropping the `typeof`

trevnorris commented on src/node_buffer.cc:
```diff
@@ -870,6 +870,66 @@ void ByteLengthUtf8(const FunctionCallbackInfo<Value> &args) {
+  else if (target_start >= target_end)
+    return args.GetReturnValue().Set(1);
+
+  uint32_t to_cmp = MIN(MIN(source_end - source_start,
```
Can we switch this to `size_t`? All your inputs are `size_t` and `memcmp` expects `size_t`.

jasnell replied:
+1

trevnorris commented:
left couple comments.

jasnell commented:
Awesome, thank you. Will update later on today

jasnell commented:
@trevnorris ... PTAL!

trevnorris commented on src/node_buffer.cc:
```diff
@@ -870,6 +870,66 @@ void ByteLengthUtf8(const FunctionCallbackInfo<Value> &args) {
+  else if (source_start >= source_end)
+    return args.GetReturnValue().Set(-1);
+  else if (target_start >= target_end)
+    return args.GetReturnValue().Set(1);
```
These should have been handled in JS. Instead can they all be turned into `CHECK`s? e.g.
```cpp
CHECK(target_start >= target_end && source_start >= source_end);
CHECK_GE(source_start >= source_end);
CHECK_GE(target_start >= target_end);
```
Since if one of these is hit then it would indicate a bug on the JS side.

trevnorris commented:
One comment, but LGTM otherwise.

jasnell commented:
@trevnorris ... updated! squashed the commits. New CI: https://ci.nodejs.org/job/node-test-pull-request/2227/

trevnorris commented:
If CI is happy LGTM

jasnell commented:
Failures in CI are unrelated.

jasnell commented:
Landed in a246689. Thank you!

jorangreef commented:
Thanks @jasnell and @trevnorris.
